Disable load-balance on servers by default

sumwale opened the pull request:
Previous set of changes checked if the host name specified in JDBC URL is a locator or server type.
This matching would fail for cases where the host in JDBC URL is different from the
hostname-for-clients setting on the servers (e.g. when using internal IP in an AWS cluster that has
public IP in hostname-for-clients).

This change adds a new flag "isCurrent" in HostAddress that marks the connected server among the
received list. The flag is then used to determine whether that member is a server or locator and
load-balance is set accordingly.

## Changes proposed in this pull request

- added "isCurrent" flag to HostAddress and re-generated thrift for java and C++ client code
- mark isCurrent flag for current member when retrieving HostAddresses in GfxdDistributionAdvisor
- use the above flag to search for current HostAddress and check its type in ClientService to
   determine the default value of load-balance
- change THashSet to UnifiedSet in GfxdDistributionAdvisor

## Patch testing

precheckin -Pstore

## Is precheckin with -Pstore clean?

Known intermittent failures

## ReleaseNotes changes

document that load-balance is false by default for servers and true for locators

## Other PRs 

NA

Commit:
Set load-balance=false on servers by default

Previous set of changes checked if the host name specified in JDBC URL is a locator or server type.
This matching would fail for cases where the host in JDBC URL is different from the hostname-for-clients
setting on the servers (e.g. when using internal IP in an AWS cluster that has public IP in hostname-for-clients).

This change adds a new flag "isCurrent" in HostAddress that marks the connected server among the received list.
The flag is then used to determine whether that member is a server or locator and load-balance is set accordingly.

kneeraj commented on gemfirexd/core/src/main/java/com/pivotal/gemfirexd/internal/engine/distributed/GfxdDistributionAdvisor.java:
```diff
@@ -337,7 +337,8 @@ else if (p instanceof BridgeServerProfile) {
             isAdded = addMemberServer(m,
                 ThriftUtils.getHostAddress(bp.getHost(), bp.getPort())
-                    .setServerType(serverType), this.thriftServers, true);
+                    .setServerType(serverType).setIsCurrent(m.equals(GemFireStore.getMyId())),
```
So the client which gets the hostaddress of the server at the connection time finds out whether it is connected to a server member or a locator member and based on that decides whether it has to do failover or not. Right?
So why can't on the server side we can just set this based on the vmkind information only?
Actually please explain when will this profile exchange happen and how setting the isCurrent on server hostaddress during profile exchange will send this information to the client side.

sumwale replied:
Server-side doesn't process this property. All handling of load-balance is done by client driver only which will decide whether to setup separate "control connection" to locator, and "data connections" to servers, use the getPreferredServer API, handle failure exceptions when appropriate to choose a new server using getPreferredServer on the "control connection" etc. Setting this property on the server-side will serve no purpose.

kneeraj replied:
Oh yes. Got confused. Thanks.

Commit:
some cleanups for load-balance setting

- search for the current host in the full host list instead of assuming that
  the one marked "current" in the list is the same (since other connections
      in the same JVM can use different URLs)
- don't try to set "implicit" load-balance again if already determined
- added a console logging properties file for tests; will be explicitly
  loaded by test logger for only the master JVM since otherwise the
  child VMs configure their own test-specific log-files
